docs/debugging.md

```diff
@@ -0,0 +1,138 @@
+For each Class and Method, you should find:
+```
+  [Class].[Method].ast.txt
+  [Class].[Method].resolved-ast.txt
+  [Class].[Method].elaborated-ast.txt
+  [Class].[Method].wf-types.txt
+  [Class].[Method].ir.txt
+  [Class].[Method].liveness.txt
+  [Class].[Method].constraints.txt
+  [Class].[Method].infer.txt
+  [Class].[Method].solver.txt
+  [Class].[Method].substitution.txt
+  [Class].[Method].types.txt
+  [Class].[Method].typed-ir.txt
+  [Class].[Method].make-region-graph.txt
+  [Class].[Method].region-graph.txt
+```
+The order is the order of the phases in the compiler, and can be used to track where something has gone wrong.
+
+Individual results can be printed to the standard output instead by passing the --print= flag to the compiler.
+For example, the following command will print the intermediate representation for the Main.main method.
+
+```
+$ veronac bank1.verona --print=Main.main.ir
+IR for Main.main:
+  Basic block BB0:
+    0 <- static Main
+    1 <- integer 100
+    2 <- call 0.example()
+[...]
+```
+
+### Debugging the parser
+
+The Verona parser currently uses [Pegmatite](https://github.com/CompilerTeaching/Pegmatite), a PEG parser design the teaching and rapid prototyping.
+This has two compile options that can aid debugging:
+
+* BUILD_TRACING  - this outputs tracing during parsing the string
+* BUILD_AST_TRACING  - this outputs the construction of the AST nodes
+
+These can be set by passing cmake `-DBUILD_TRACING=On` and `-DBUILD_AST_TRACING=On` respectively.
+
+## Debugging the interpreter
+
+As the majority of code running in the interpreter is actually the runtime, the subsequent section could also prove useful.
+The interpreter provides a trace of all the instructions that it executes.
+Every option that you can pass to the interpreter, can also be passed to the compiler prepended with `run-`, for example:
+```
+interpreter foo.bc --verbose
+```
+or 
+```
+veronac foo.verona --run --run-verbose
+```
+The `--verbose` option basically provides a trace of the interpreters operations:
+for example:
+```
+veronac --run --run-verbose region101.verona
+```
+produces
+```
+[ 27b]: Entering function Main.main, argc=1 retc=1 locals=9
+[ 28d]: LOAD_DESCRIPTOR r1, 0x1
+[ 293]: COPY r8, r1
+[ 296]: CALL 7, 0x1
+[ 296]:  Calling function Main.test1, base=r8 argc=1 retc=1 locals=37
+[ 12c]:  LOAD_DESCRIPTOR r1, 0x3
+...
+```
+
+Both the interpreter, and the compiler's run option, can also provide systematic testing for concurrency:
+```
+  veronac-sys --run testsuite/demo/run-pass/dining_phil.verona --run-seed 100 --run-seed_upper 200
+```
+runs 100 seeds sequentially, but testing various interleavings of the runtime. The seeds are replayable.
+
+## Debugging the runtime
+
+The runtime provides two key features to aid debugging:
+
+* Systematic testing
+* Crash logging (A flight recorder that provides the last few seconds of logging)
+
+Both of these can produce high detail logging of the runtime operations.
```

Comment:
This would be a lot of work, but walking through an example of debugging a crash in systematic testing would be super helpful. Or at the very least, showing a few lines of systematic output and explaining the convention (the "prefix" numbers represent threads, each object prints its address, etc.).

Reply:
I'll add a snippet, not sure I know how to explain that until I get some questions from people.